components/admin/QuestDetails.tsx

```diff
@@ -0,0 +1,214 @@
+      if ((data as UserTask[]).length) setTasks(data as UserTask[]);
+    });
+  }, [questId]);
+
+  const generateOAuthUrl = (task: UserTask): string => {
+    if (!address) {
+      showNotification("Please connect wallet to test", "info");
+      return "";
+    }
+    if (task.verify_endpoint_type === "oauth_discord") {
+      const rootUrl = "https://discord.com/api/oauth2/authorize";
+      const options = {
+        redirect_uri: `${process.env.NEXT_PUBLIC_API_LINK}/${task.verify_endpoint}`,
+        client_id: process.env.NEXT_PUBLIC_DISCORD_CLIENT_ID as string,
+        response_type: "code",
+        scope: ["identify", "guilds"].join(" "),
+        state: `${hexToDecimal(address)}+${task.quest_id}+${task.id}`,
+      };
+      const qs = new URLSearchParams(options).toString();
+      return `${rootUrl}?${qs}`;
+    } else {
+      const codeChallenge = generateCodeChallenge(
+        process.env.NEXT_PUBLIC_TWITTER_CODE_VERIFIER as string
+      );
+      const rootUrl = "https://twitter.com/i/oauth2/authorize";
+      const options = {
+        redirect_uri: `${task.verify_endpoint}?addr=${hexToDecimal(address)}`,
+        client_id: process.env.NEXT_PUBLIC_TWITTER_CLIENT_ID as string,
+        state: "state",
+        response_type: "code",
+        code_challenge: codeChallenge,
+        code_challenge_method: "S256",
+        scope: ["follows.read", "tweet.read", "users.read"].join(" "),
+      };
+      const qs = new URLSearchParams(options).toString();
+      return `${rootUrl}?${qs}`;
+    }
+  };
+
+  useEffect(() => {
+    // get `error_msg` from url
+    if (typeof window !== "undefined") {
+      const url = window.location;
+      // Your client-side code that uses window goes here
+      const urlParams = new URLSearchParams(url.search);
+      const error_msg = urlParams.get("error_msg");
+      if (error_msg) {
+        setCustomError(error_msg);
+      }
+    }
+  }, []);
+
+  return (
+    <>
+      <NftImage
+        nfts={quest?.rewards_nfts?.map((nft: NFTItem) => {
+          return { imgSrc: nft.img, level: nft.level };
+        })}
+      />
+
+      <div className={styles.descriptionContainer}>
+        {quest.name === "loading" ? (
+          <Skeleton
+            variant="text"
+            width={400}
+            sx={{ fontSize: "2rem", bgcolor: "grey.900" }}
+          />
+        ) : (
+          <h1 className="title extrabold mt-5 mw-90">{quest.name}</h1>
+        )}
+        {quest.desc === "loading" ? (
+          <Skeleton
+            variant="text"
+            width={350}
+            sx={{ fontSize: "0.8rem", bgcolor: "grey.900" }}
+          />
+        ) : (
+          <>
+            <p className="text-center max-w-[50vw]">{quest.desc}</p>
+            {quest?.additional_desc ? (
+              <>
+                <p className="text-center max-w-[50vw]">
+                  {quest.additional_desc}
+                </p>
+              </>
+            ) : null}
+          </>
+        )}
```

Comment:
Couldn't we do a loading component here ? Seems again to be a long component that we can subdivize

Reply:
done